libgadget/fof.c

```diff
@@ -401,13 +378,25 @@ void fof_label_primary(ForceTree * tree, MPI_Comm Comm)
         treewalk_run(tw, NULL, PartManager->NumPart);
 
         t1 = second();
-
+        /* Just in case we missed some during HEAD building.*/
+        for(i = 0; i < PartManager->NumPart; i++) {
+            int head = HEAD(i, FOF_PRIMARY_GET_PRIV(tw)->Head);
+            if(HaloLabel[head].MinID > HaloLabel[i].MinID) {
+                HaloLabel[head].MinID = HaloLabel[i].MinID;
+                HaloLabel[head].MinIDTask = HaloLabel[i].MinIDTask;
+            }
+        }
         /* let's check out which particles have changed their MinID,
          * mark them for next round. */
         link_across = 0;
 #pragma omp parallel for reduction(+: link_across)
         for(i = 0; i < PartManager->NumPart; i++) {
-            MyIDType newMinID = HaloLabel[HEAD(i, tw)].MinID;
+            int head = HEAD(i, FOF_PRIMARY_GET_PRIV(tw)->Head);
+            if(i != head) {
```

Comment:
This loop is also changing the MinID -- but the comment above still says it is only checking changed particles.

What is the difference between this minid assignment and the 'just in case' above? Looks like they are just overwriting each other. Can they be merged?

Reply:
The difference is that the above case sets the MinID of the HEAD to min(ID of all particles). This sets the ID of the particles to the minID of the HEAD. So they must be separate loops.

Reply:
Thanks. I see.